pulsar-metadata/src/main/java/org/apache/pulsar/metadata/tableview/impl/MetadataStoreTableViewImpl.java

```diff
@@ -0,0 +1,329 @@
+            return null;
+        });
+    }
+
+    private void handleNotification(org.apache.pulsar.metadata.api.Notification notification) {
+
+        if (notification.getType() == NotificationType.ChildrenChanged) {
+            return;
+        }
+
+        String path = notification.getPath();
+
+        doHandleNotification(path);
+    }
+
+
+    private CompletableFuture<Void> handleExisting(String path) {
+        if (!isValidPath(path)) {
+            return CompletableFuture.completedFuture(null);
+        }
+        return cache.get(path)
+                .thenAccept(valOpt -> {
+                    valOpt.ifPresent(val -> {
+                        String key = getKey(path);
+                        updateData(key, val);
+                        if (log.isDebugEnabled()) {
+                            log.debug("{} applying existing item key={} value={}",
+                                    name,
+                                    key,
+                                    val);
+                        }
+                        for (var listener : existingItemListeners) {
+                            try {
+                                listener.accept(key, val);
+                            } catch (Throwable e) {
+                                log.error("{} failed to listen existing item key:{}, val:{}", name, key, val,
+                                        e);
+                                throw e;
+                            }
+                        }
+                    });
+                });
+    }
+
+    private void fill() throws MetadataStoreException {
+        log.info("{} start filling existing items under the pathPrefix:{}", name, pathPrefix);
+        ConcurrentLinkedDeque<String> q = new ConcurrentLinkedDeque<>();
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+        q.add(pathPrefix);
+        LongAdder count = new LongAdder();
+        while (!q.isEmpty()) {
+            int size = Math.min(MAX_CONCURRENT_METADATA_OPS_DURING_FILL, q.size());
+            for (int i = 0; i < size; i++) {
+                String path = q.poll();
+                futures.add(store.getChildren(path)
+                        .thenCompose(children -> {
+                            // The path is leaf
+                            if (children.isEmpty()) {
+                                count.increment();
+                                return handleExisting(path);
+                            } else {
+                                for (var child : children) {
+                                    q.add(path + "/" + child);
+                                }
+                                return CompletableFuture.completedFuture(null);
+                            }
+                        }));
+            }
+            try {
+                FutureUtil.waitForAll(futures).get(timeoutInMillis * size, TimeUnit.MILLISECONDS);
```

Comment:
This timeout might be too long if the size of children is too large (at most 50*30s = 25 minutes).

Should we just use `timeoutInMillis` instead? Also, I'm wondering if we can ignore the exception (with just a warning log) and just skip the loop in this case. Because `fill()` should not be required. It's just used to establish an initial cache view.

Reply:
I think we can give 5 mins at most.

I am worried about the worst case with really large table view (although this table view should be relatively small)

Broker would behave wrong if any listener logic fails during this fill.

It is also important to start with complete view for orphan ownership cleanup - the leader needs to scan all current ownerships and reassign orphan ones asap upon any unexpected broker crash.

Reply:
> with really large table view

It should be noted that this timeout is applied in **each loop**. 5 minutes are still large for at most 50 concurrent metadata store operations.

Reply:
Should we use 5 minutes as the total timeout in the loop?

Reply:
ok total 5mins timeout make sense.